initial vmq_diversity auth_cache

dergraf opened the pull request.

larshesel committed:
Fix multiple bridge bug

It's unclear for what the `lists:usort/1` was used, but it's certainly a
bug to change the order of the results and prevent multiple bridges from
being defined.

dergraf committed:
initial vmq_diversity auth_cache

dergraf committed:
Lua Postgres Authentication/Authorization Plugin

dergraf committed:
Lua MySQL Authentication/Authorization Plugin

dergraf committed:
Lua MongoDB Authentication/Authorization Plugin

This plugin uses bcrypt to hash and store the password. It relies on the
following commit that introduces the newest Rebar3

dergraf committed:
Upgraded Rebar3

dergraf committed:
wip, redis auth plugin

dergraf committed:
Merge pull request #270 from larshesel/the-big-integration

The big integration

dergraf committed:
upgraded rebar3, required to upgrade node_package dependency

larshesel committed:
Support vmq-admin usage lead lines for plugins

larshesel committed:
Add webhooks cache (#273)

larshesel commented on apps/vmq_diversity/priv/auth/mysql.lua:
```diff
@@ -0,0 +1,119 @@
+--    password, publish_acl, subscribe_acl)
+-- VALUES 
+--   ('', 'test-client', 'test-user', 
+--    PASSWORD('123'), '["a/b/c","c/b/#"]', '["a/b/c","c/b/#"]');
```
The MySQL `PASSWORD()` function stores unsalted passwords. Shall we just accept that for MySQL as that's probably what people need or should we go for something more modern?

larshesel commented on apps/vmq_diversity/priv/vmq_diversity.schema:
```diff
@@ -20,11 +20,31 @@
+%%   vmq_diversity.mongodb.file = {{luascripts_dir}}/auth/mongodb.lua
+%%
+%% For Redis:
+%%   vmq_diversity.redis.file = {{luascripts_dir}}/auth/redis.lua
```
The `luascripts_dir` will by default point to `share/lua` and the scripts you mention won't be there I believe. So maybe they should be copied there in the vernemq.conf. Is it confusing that we say that anything in `luascripts_dir` will be automatically loaded and then have specific, non-autoload files in the `luascripts_dir/auth` folder? Maybe we should change the structure to `luascripts_dir/autoload` and have the non-autoload scripts in the `luascripts_dir` folder?

larshesel commented on apps/vmq_diversity/src/vmq_diversity_plugin.erl:
```diff
@@ -200,21 +200,47 @@ auth_on_register(Peer, SubscriberId, UserName, Password, CleanSession) ->
-                                  {topic, unword(Topic)},
-                                  {payload, Payload},
-                                  {retain, IsRetain}]).
+    case vmq_diversity_cache:match_publish_acl(MP, ClientId, Topic) of
```
I think qos and retain should also be part of the match. Otherwise you can't prevent the client from publishing retained messages when using the cache, even if the client is not allowed to do so by the hook logic. Also, I think it should be possible to allow only for specific QoS levels. Of course this is the way I did it in vmq_webhooks and it needs to be optimized so QoS levels and retained bits get's merged into one entry to prevent combinatorial cache size explosions.

larshesel reviewed (commented):
Looks good - mostly concerned about the cache retained/qos stuff

larshesel committed:
Fix dev build targets

dergraf committed:
Merge pull request #277 from larshesel/fix-dev-build-targets

Fix dev build targets

dergraf committed:
Merge pull request #276 from larshesel/fix-vmq-admin-usage-for-plugins

It's a bit hacky (especially because of the required spacing, but works for now)
Support vmq-admin usage lead lines for plugins

dergraf committed:
Merge pull request #271 from larshesel/fix-multiple-bridge-issue

Fix multiple bridge issue

dergraf committed:
🎂 It's my birthday, give me cake! Upgraded node_package

dergraf committed:
Merge pull request #274 from erlio/upgrade_rebar3

upgraded rebar3, required to upgrade node_package dependency
peer reviewed with @larshesel

larshesel committed:
Add bridge configuration changes to the changelog

larshesel committed:
Run all tests on travis (#281)

dergraf committed:
Support modifiers in ACL-cache

dergraf committed:
several fixes, validate modifiers, use new JSON object based ACLs

this commit contains a changes the return value for json:decode/encode
to nil in case of a decoding/encoding error happens. Previously we
returned 'false' which doesn't make sense.

dergraf committed:
wip

dergraf committed:
remove database configurations from auth_cache Lua scripts

dergraf committed:
some refactorings, return modifiers in auth_on_subscribe/on_unsubscribe

larshesel commented:
I think we can close this one as i expect the Pr of the branch against master to be the relevant one.
